From b8287db0da577d4ff540fb058d2868cddacfc6ee Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 30 Mar 2022 10:18:15 +0300
Subject: [PATCH] [SW-66675] rom: tpm: add tpm startup routine

This patch ports the tpm startup routine to rom from preboot.
TPM_startup() sets up the tpm slave to a command ready state
that would enable us to instruct tpm with extend command.

Change-Id: I413918206bd044573ce435b565276189cedf4e6e
---

diff --git a/arch/gaudi3/arc/tpm2.c b/arch/gaudi3/arc/tpm2.c
index 903e320..86cae1f 100644
--- a/arch/gaudi3/arc/tpm2.c
+++ b/arch/gaudi3/arc/tpm2.c
@@ -12,6 +12,8 @@
 #include <errno.h>
 #include <dev_common.h>
 #include <linux/types.h>
+#include <tpm2.h>
+#include <hl_rsa_arc.h>
 
 /* Internal error of TPM command library */
 #define TPM_LIB_ERROR ((uint32_t)~0u)
@@ -19,6 +21,8 @@
 
 static struct spi_ctx tpm_spi = {0};
 static int locality;
+uint8_t cmd_response_buffer[COMMAND_BUFFER_SIZE];
+uint8_t tpm_buf[TPM_DEV_BUFSIZE + sizeof(uint8_t)];  /* Max buffer size + addr */
 
 static unsigned int get_cs_pad_sel_tpm(void)
 {
@@ -214,18 +218,94 @@
 	return tpm_tis_spi_xfer(addr, out, NULL, len);
 }
 
-static int tpm_tis_wait_init(int loc)
+static int tpm_tis_spi_read32(uint32_t addr, uint32_t *result)
 {
-	uint8_t status;
+	__le32 result_le;
 	int ret;
-	uint32_t count = TIS_LONG_TIMEOUT_MS / TPM_TIMEOUT_MS;
+
+	ret = tpm_tis_spi_read(addr, (uint8_t *)&result_le, sizeof(uint32_t));
+	if (!ret)
+		*result = le32_to_cpu(result_le);
+
+	return ret;
+}
+
+static int tpm_command_size(const void *command)
+{
+	const size_t command_size_offset = 2;
+
+	/* get unaligned be32 */
+	return get_unaligned_be32((__be32 *)(command + command_size_offset));
+}
+
+static uint32_t tpm_return_code(const void *response)
+{
+	const size_t return_code_offset = 6;
+
+	return get_unaligned_be32(response + return_code_offset);
+}
+
+static uint8_t tpm_spi_status(uint8_t *status)
+{
+	return tpm_tis_spi_read(TPM_STS(locality), status, 1);
+}
+
+static int tpm_tis_spi_cancel(void)
+{
+	uint8_t data = TPM_STS_COMMAND_READY;
+
+	return tpm_tis_spi_write(TPM_STS(locality), &data, 1);
+}
+
+static void tpm_tis_spi_release_locality(int loc, bool force)
+{
+	const uint8_t mask = TPM_ACCESS_REQUEST_PENDING | TPM_ACCESS_VALID;
+	uint8_t buf;
+
+	if (tpm_tis_spi_read(TPM_ACCESS(loc), &buf, 1) < 0)
+		return;
+
+	if (force || (buf & mask) == mask) {
+		buf = TPM_ACCESS_ACTIVE_LOCALITY;
+		tpm_tis_spi_write(TPM_ACCESS(loc), &buf, 1);
+	}
+}
+
+static int tpm_spi_get_burstcount(void)
+{
+	uint32_t burstcount, ret;
+	uint32_t count;
+
+	/* wait for burstcount */
+	count = TIS_SHORT_TIMEOUT_MS / TPM_TIMEOUT_MS;
+	do {
+		ret = tpm_tis_spi_read32(TPM_STS(locality),
+					 &burstcount);
+		if (ret)
+			return -1;
+
+		burstcount = (burstcount >> 8) & 0xFFFF;
+		if (burstcount)
+			return burstcount;
+
+		udelay(MS_TO_US(TPM_TIMEOUT_MS));
+		count --;
+	} while (count > 0);
+
+	return -1;
+}
+
+static int tpm_spi_wait_for_stat(uint8_t mask, unsigned long timeout, uint8_t *status)
+{
+	uint32_t count = timeout / TPM_TIMEOUT_MS;
+	int ret;
 
 	do {
-		ret = tpm_tis_spi_read(TPM_ACCESS(loc), &status, 1);
+		ret = tpm_spi_status(status);
 		if (ret)
-			break;
+			return ret;
 
-		if (status & TPM_ACCESS_VALID)
+		if ((*status & mask) == mask)
 			return 0;
 
 		udelay(MS_TO_US(TPM_TIMEOUT_MS));
@@ -233,6 +313,37 @@
 	} while (count > 0);
 
 	return -1;
+}
+
+static uint8_t tpm_tis_spi_valid_status(uint8_t *status)
+{
+	return tpm_spi_wait_for_stat(TPM_STS_VALID, TIS_SHORT_TIMEOUT_MS,
+					status);
+}
+
+static int tpm_spi_recv_data(uint8_t *buf, size_t count)
+{
+	size_t size = 0;
+	int burstcnt, len, ret;
+	uint8_t status;
+
+	while (size < count &&
+	       tpm_spi_wait_for_stat(TPM_STS_DATA_AVAIL | TPM_STS_VALID,
+					 TIS_SHORT_TIMEOUT_MS, &status) == 0) {
+		burstcnt = tpm_spi_get_burstcount();
+		if (burstcnt < 0)
+			return burstcnt;
+
+		len = min_t(int, burstcnt, count - size);
+		ret = tpm_tis_spi_read(TPM_DATA_FIFO(locality),
+				       buf + size, len);
+		if (ret)
+			return ret;
+
+		size += len;
+	}
+
+	return size;
 }
 
 static int tpm_tis_spi_check_locality(int loc)
@@ -288,16 +399,147 @@
 	return ret;
 }
 
-static int tpm_tis_spi_read32(uint32_t addr, uint32_t *result)
+static int tpm_spi_recv(uint8_t *buf, size_t count)
 {
-	__le32 result_le;
-	int ret;
+	size_t size, expected;
 
-	ret = tpm_tis_spi_read(addr, (uint8_t *)&result_le, sizeof(uint32_t));
-	if (!ret)
-		*result = le32_to_cpu(result_le);
+	if (count < TPM_HEADER_SIZE) {
+		size = -1;
+		goto out;
+	}
+
+	size = tpm_spi_recv_data(buf, TPM_HEADER_SIZE);
+	if (size < TPM_HEADER_SIZE) {
+		/* TODO: Check if size invalid requires to be returned */
+		goto out;
+	}
+
+	expected = get_unaligned_be32(buf + 2);
+	if (expected > count) {
+		size = -1;
+		goto out;
+	}
+
+	size += tpm_spi_recv_data(&buf[TPM_HEADER_SIZE],
+				   expected - TPM_HEADER_SIZE);
+	if (size < expected) {
+		size = -1;
+		goto out;
+	}
+
+out:
+	tpm_tis_spi_cancel();
+	tpm_tis_spi_release_locality(locality, false);
+
+	return size;
+}
+
+static int tpm_spi_send(const uint8_t *buf, size_t len)
+{
+	uint32_t i, size;
+	uint8_t status;
+	int burstcnt, ret;
+	uint8_t data;
+
+	if (len > TPM_DEV_BUFSIZE)
+		return -E2BIG;  /* Command is too long for our tpm, sorry */
+
+	ret = tpm_tis_spi_request_locality(0);
+	if (ret < 0)
+		return -EBUSY;
+
+	/*
+	 * Check if the TPM is ready. If not, if not, cancel the pending command
+	 * and poll on the status to be finally ready.
+	 */
+	ret = tpm_spi_status(&status);
+	if (ret)
+		return ret;
+
+	if (!(status & TPM_STS_COMMAND_READY)) {
+		/* Force the transition, usually this will be done at startup */
+		ret = tpm_tis_spi_cancel();
+		if (ret) {
+			goto out_err;
+		}
+
+		ret = tpm_spi_wait_for_stat(TPM_STS_COMMAND_READY,
+						TIS_LONG_TIMEOUT_MS, &status);
+		if (ret < 0 || !(status & TPM_STS_COMMAND_READY)) {
+			goto out_err;
+		}
+	}
+
+	for (i = 0; i < len - 1;) {
+		burstcnt = tpm_spi_get_burstcount();
+		if (burstcnt < 0)
+			return burstcnt;
+
+		size = min_t(int, len - i - 1, burstcnt);
+		ret = tpm_tis_spi_write(TPM_DATA_FIFO(locality),
+					buf + i, size);
+		if (ret < 0)
+			goto out_err;
+
+		i += size;
+	}
+
+	ret = tpm_tis_spi_valid_status(&status);
+	if (ret)
+		goto out_err;
+
+	if ((status & TPM_STS_DATA_EXPECT) == 0) {
+		ret = -EIO;
+		goto out_err;
+	}
+
+	ret = tpm_tis_spi_write(TPM_DATA_FIFO(locality),
+				buf + len - 1, 1);
+	if (ret)
+		goto out_err;
+
+	ret = tpm_tis_spi_valid_status(&status);
+	if (ret)
+		goto out_err;
+
+	if ((status & TPM_STS_DATA_EXPECT) != 0) {
+		ret = -EIO;
+		goto out_err;
+	}
+
+	data = TPM_STS_GO;
+	ret = tpm_tis_spi_write(TPM_STS(locality), &data, 1);
+	if (ret)
+		goto out_err;
+
+	return len;
+
+out_err:
+	tpm_tis_spi_cancel();
+	tpm_tis_spi_release_locality(locality, false);
 
 	return ret;
+}
+
+static int tpm_tis_wait_init(int loc)
+{
+	uint8_t status;
+	int ret;
+	uint32_t count = TIS_LONG_TIMEOUT_MS / TPM_TIMEOUT_MS;
+
+	do {
+		ret = tpm_tis_spi_read(TPM_ACCESS(loc), &status, 1);
+		if (ret)
+			break;
+
+		if (status & TPM_ACCESS_VALID)
+			return 0;
+
+		udelay(MS_TO_US(TPM_TIMEOUT_MS));
+		count--;
+	} while (count > 0);
+
+	return -1;
 }
 
 int tpm2_init(void)
@@ -331,3 +573,82 @@
 
 	return 0;
 }
+
+static int tpm_xfer(const uint8_t *sendbuf, size_t send_size,
+	uint8_t *recvbuf, size_t *recv_size)
+{
+	unsigned int count;
+	int ret = 0;
+
+	/* switch endianness: big->little */
+	count = get_unaligned_be32(sendbuf + TPM_CMD_COUNT_BYTE);
+
+	if (count == 0) {
+		return -1;
+	}
+	if (count > send_size) {
+		return -1;
+	}
+
+	ret = tpm_spi_send(sendbuf, send_size);
+	if (ret < 0)
+		return ret;
+
+	ret = tpm_spi_recv(tpm_buf, sizeof(tpm_buf));
+	if (ret >= 0) {
+		if (ret > (int)*recv_size)
+			return -1;
+		memcpy(recvbuf, tpm_buf, ret);
+		*recv_size = ret;
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static int tpm_spi_sendrecv_command(const void *command,
+			 void *response, size_t *size_ptr)
+{
+	int err;
+	size_t response_length;
+
+	if (response) {
+		response_length = *size_ptr;
+	} else {
+		response = cmd_response_buffer;
+		response_length = sizeof(cmd_response_buffer);
+	}
+
+	err = tpm_xfer(command, tpm_command_size(command),
+		       response, &response_length);
+
+	if (err < 0)
+		return err;
+
+	if (size_ptr)
+		*size_ptr = response_length;
+
+	return tpm_return_code(response);
+}
+
+#define TPM_STARTUP_COMMAND_BUF_SIZE	12
+int tpm2_startup(enum tpm2_startup_types mode)
+{
+	const uint8_t command_v2[TPM_STARTUP_COMMAND_BUF_SIZE] = {
+		tpm_u16(TPM2_ST_NO_SESSIONS),
+		tpm_u32(TPM_STARTUP_COMMAND_BUF_SIZE),
+		tpm_u32(TPM2_CC_STARTUP),
+		tpm_u16(mode),
+	};
+	int ret;
+
+	/*
+	 * Note TPM2_Startup command will return RC_SUCCESS the first time,
+	 * but will return RC_INITIALIZE otherwise.
+	 */
+	ret = tpm_spi_sendrecv_command(command_v2, NULL, NULL);
+	if (ret && ret != TPM2_RC_INITIALIZE)
+		return ret;
+
+	return 0;
+}
